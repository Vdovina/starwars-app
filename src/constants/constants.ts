export const PAGE_SIZE = 10;
export const DEFAULT_PAGES = [0, 1];
export const VIRTUALIZATION_DELAY = 2000;
