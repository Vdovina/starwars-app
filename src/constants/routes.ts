const SERVER = 'https://swapi.dev/api';

export const ROUTES = {
  HOME: '/',
  FILMS: '/films',
  FILM_CARD: '/films/:id',
  CHARACTERS: '/characters',
  CHARACTER_CARD: '/characters/:id',
  STARSHIPS: '/starships',
  STARSHIP_CARD: '/starships/:id',
  PLANETS: '/planets',
  PLANET_CARD: '/planets/:id',
  SPECIES: '/species',
  SPECIES_CARD: '/species/:id',
  VEHICLES: '/vehicles',
  VEHICLE_CARD: '/vehicles/:id'
};

export const API_ROUTES = {
  GET_FILMS: `${SERVER}/films`,
  GET_FILM_INFO: `${SERVER}/films/:id`,
  GET_CHARACTERS: `${SERVER}/people`,
  GET_CHARACTER_INFO: `${SERVER}/people/:id`,
  GET_STARSHIPS: `${SERVER}/starships`,
  GET_STARSHIP_INFO: `${SERVER}/starships/:id`,
  GET_PLANETS: `${SERVER}/planets`,
  GET_PLANET_INFO: `${SERVER}/planets/:id`,
  GET_SPECIES: `${SERVER}/species`,
  GET_SPECIES_INFO: `${SERVER}/species/:id`,
  GET_VEHICLES: `${SERVER}/vehicles`,
  GET_VEHICLE_INFO: `${SERVER}/vehicles/:id`,
};
