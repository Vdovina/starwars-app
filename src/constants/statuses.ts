export enum STATUSES {
  LOADING = 'loading',
  ERROR = 'hasError',
  SUCCESS = 'hasValue',
}
