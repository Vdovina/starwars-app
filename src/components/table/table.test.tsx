import { create, act as testAct, ReactTestRenderer } from 'react-test-renderer';
import { render as testRender, fireEvent } from '@testing-library/react';
import { TColumn, TRow, Table } from './';
import { LoadingStatus } from '../../constants/statuses';

interface TestTableData {
  col1: string;
  col2: string;
}

const columns: TColumn<TestTableData>[] = [
  {
    accessor: 'col1',
    Header: 'Col1',
  },
  {
    accessor: 'col2',
    Header: 'Col2',
  },
];
const data: TRow<TestTableData>[] = [
  {
    col1: 'Test1-1',
    col2: 'Test1-2',
    name: '1',
    url: 'some-path/1/',
  },
  {
    col1: 'Test2-1',
    col2: 'Test2-2',
    name: '2',
    url: 'some-path/2/',
  },
];

describe('Table Tests', () => {
  it('render empty table', () => {
    let component: ReactTestRenderer;
    testAct(() => {
      component = create(<Table rows={[]} columns={[]} total={0} />);
    });
    expect(component.toJSON()).toMatchSnapshot();
  });

  it('render table with data', () => {
    let component: ReactTestRenderer;
    testAct(() => {
      component = create(<Table rows={data} columns={columns} total={data.length} />);
    });
    expect(component.toJSON()).toMatchSnapshot();
  });

  it('render table with loading', () => {
    let component: ReactTestRenderer;
    testAct(() => {
      component = create(<Table rows={[]} columns={[]} total={0} loading={LoadingStatus.Loading} />);
    });
    expect(component.toJSON()).toMatchSnapshot();
  });

  it('render table with error', () => {
    let component: ReactTestRenderer;
    testAct(() => {
      component = create(<Table rows={[]} columns={[]} total={0} loading={LoadingStatus.Error} />);
    });
    expect(component.toJSON()).toMatchSnapshot();
  });

  it('on row click', () => {
    const onClick = jest.fn();
    const { container } = testRender(<Table rows={data} columns={columns} total={data.length} onRowClick={onClick} />);
    const row = container.querySelectorAll<HTMLTableRowElement>('tbody tr')[0] as HTMLTableRowElement;
    fireEvent.click(row);
    expect(onClick).toHaveBeenCalledWith('1');
  });

  it('on row click without callback', () => {
    const onClick = jest.fn();
    const { container } = testRender(<Table rows={data} columns={columns} total={data.length} />);
    const row = container.querySelectorAll<HTMLTableRowElement>('tbody tr')[0] as HTMLTableRowElement;
    fireEvent.click(row);
    expect(onClick).not.toHaveBeenCalled();
  });

  it('on scroll', () => {
    const onScroll = jest.fn();
    let component: ReactTestRenderer;
    testAct(() => {
      component = create(
        <div style={{ height: 100 }}>
          <Table rows={data} columns={columns} total={20} size={5} onLoad={onScroll} />
        </div>
      );
    });
    const tbody = component.root.findByType('tbody');
    testAct(() => tbody.props.onScroll({ currentTarget: { scrollTop: 100 } }));
    expect(onScroll).toHaveBeenCalledWith([1]);
  });

  it('on scroll longer page', () => {
    const onScroll = jest.fn();
    let component: ReactTestRenderer;
    testAct(() => {
      component = create(
        <div style={{ height: 100 }}>
          <Table rows={data} columns={columns} total={20} size={10} onLoad={onScroll} />
        </div>
      );
    });
    const tbody = component.root.findByType('tbody');
    testAct(() => tbody.props.onScroll({ currentTarget: { scrollTop: 100 } }));
    expect(onScroll).not.toHaveBeenCalled();
  });
});
