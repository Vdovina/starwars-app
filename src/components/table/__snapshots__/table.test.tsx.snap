// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table Tests render empty table 1`] = `
<div
  className="table-container"
>
  <table
    role="table"
  >
    <thead>
      <tr />
    </thead>
    <tbody
      onScroll={[Function]}
      role="rowgroup"
    >
      <tr
        className="msg-placeholder"
      >
        <td
          className="msg-error"
        >
          Information not found
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;

exports[`Table Tests render table with data 1`] = `
<div
  className="table-container"
>
  <table
    role="table"
  >
    <thead>
      <tr
        role="row"
      >
        <th
          colSpan={1}
          role="columnheader"
        >
          <div>
            Col1
          </div>
        </th>
        <th
          colSpan={1}
          role="columnheader"
        >
          <div>
            Col2
          </div>
        </th>
      </tr>
    </thead>
    <tbody
      onScroll={[Function]}
      role="rowgroup"
    >
      <tr
        onClick={[Function]}
        role="row"
      >
        <td
          role="cell"
        >
          Test1-1
        </td>
        <td
          role="cell"
        >
          Test1-2
        </td>
      </tr>
      <tr
        onClick={[Function]}
        role="row"
      >
        <td
          role="cell"
        >
          Test2-1
        </td>
        <td
          role="cell"
        >
          Test2-2
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;

exports[`Table Tests render table with error 1`] = `
<div
  className="table-container"
>
  <table
    role="table"
  >
    <thead>
      <tr />
    </thead>
    <tbody
      onScroll={[Function]}
      role="rowgroup"
    >
      <tr
        className="msg-placeholder"
      >
        <td
          className="msg-error"
        >
          Information not found
        </td>
      </tr>
      <tr
        className="msg-placeholder"
      >
        <td
          className="msg-error"
        >
          Something went wrong
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;

exports[`Table Tests render table with loading 1`] = `
<div
  className="table-container"
>
  <table
    role="table"
  >
    <thead>
      <tr />
    </thead>
    <tbody
      onScroll={[Function]}
      role="rowgroup"
    >
      <tr
        className="loader-placeholder"
      >
        <td>
          <svg
            height="24"
            viewBox="0 0 40 24"
            width="80"
            xmlns="http://www.w3.org/2000/svg"
          >
            <circle
              cx="40"
              cy="12"
              fill="currentColor"
              r="0"
            >
              <animate
                attributeName="r"
                begin=".67"
                calcMode="spline"
                dur="1.5s"
                keySplines="0.2 0.2 0.4 0.8;0.2 0.2 0.4 0.8;0.2 0.2 0.4 0.8"
                repeatCount="indefinite"
                values="0;6;0;0"
              />
            </circle>
            <circle
              cx="20"
              cy="12"
              fill="currentColor"
              r="0"
            >
              <animate
                attributeName="r"
                begin=".33"
                calcMode="spline"
                dur="1.5s"
                keySplines="0.2 0.2 0.4 0.8;0.2 0.2 0.4 0.8;0.2 0.2 0.4 0.8"
                repeatCount="indefinite"
                values="0;6;0;0"
              />
            </circle>
            <circle
              cx="0"
              cy="12"
              fill="currentColor"
              r="0"
            >
              <animate
                attributeName="r"
                begin="0"
                calcMode="spline"
                dur="1.5s"
                keySplines="0.2 0.2 0.4 0.8;0.2 0.2 0.4 0.8;0.2 0.2 0.4 0.8"
                repeatCount="indefinite"
                values="0;6;0;0"
              />
            </circle>
          </svg>
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
